components/esp_modem/include/esp_private/c_api_wrapper.hpp

```diff
@@ -0,0 +1,87 @@
+// Copyright 2022 Espressif Systems (Shanghai) PTE LTD
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @file c_api_wrapper.hpp
+ * @brief Collection of C API wrappers
+ * 
+ * This file is located in include/esp_private because it is not intended for users,
+ * but rather for esp_modem C extension developers.
+ * 
+ * The C extension API must provide a 'factory function' that returns initialized pointer to esp_modem_dce_wrap.
+ * Helper functions provided below, can be used for conversion between C++ enum classes and standard C enums.
+ */
+
+#pragma once
+
+#include "cxx_include/esp_modem_dce_factory.hpp"
+#include "esp_modem_c_api_types.h"
+
+using namespace esp_modem;
+
+struct esp_modem_dce_wrap { // need to mimic the polymorphic dispatch as CPP uses templated dispatch
+    enum class modem_wrap_dte_type { UART, VFS, USB } dte_type;
+    dce_factory::ModemType modem_type;
+    DCE *dce;
+    DTE *dte; // Plain pointer to DTE. Lifetime of the underlaying DTE object is managed by DCE.

```

Comment:
```suggestion
    std::make_shared<DTE> dte;
```

Reply:
Just note, that I was trying to keep the headers compatible with C++11, but sources could use C++17 constructs. This is okay if this header is private only.